lib/firmware/cosmos.c

```diff
@@ -0,0 +1,307 @@
+// 45 + 16 = 61
+#define SIGNING_TEMPLATE_MSG_SEND_SEG3 "%s\",\"to_address\":\""
+// 45 + 3 = 48
+#define SIGNING_TEMPLATE_MSG_SEND_SEG4 "%s\"}}"
+
+// 16 + ^20 = ^36
+#define SIGNING_TEMPLATE_END_SEG1 "],\"sequence\":\"%" PRIu64 "\"}"
+
+void sha256UpdateEscaped(SHA256_CTX *_ctx, const char *s, size_t len)
+{
+    while (len > 0)
+    {
+        if (s[0] == '"')
+        {
+            sha256_Update(_ctx, (uint8_t *)"\\\"", 2);
+        }
+        else if (s[0] == '\\')
+        {
+            sha256_Update(_ctx, (uint8_t *)"\\\\", 2);
+        }
+        else
+        {
+            sha256_Update(_ctx, (uint8_t *)&s[0], 1);
+        }
+        s = &s[1];
+        len--;
+    }
+}
+
+bool cosmos_signTxInit(const HDNode* _node,
+                       const uint32_t* _address_n,
+                       const size_t _address_n_count,
+                       const uint64_t account_number,
+                       const char *chain_id,
+                       const size_t chain_id_length,
+                       const uint32_t fee_uatom_amount,
+                       const uint32_t gas,
+                       const char *memo,
+                       const size_t memo_length,
+                       const uint64_t _sequence,
+                       const uint32_t msg_count)
+{
+
+    initialized = true;
+    msgs_remaining = msg_count;
+    has_message = false;
+    sequence = _sequence;
+
+    memzero(&node, sizeof(node));
+    memcpy(&node, _node, sizeof(node));
+    memzero(address_n, sizeof(address_n));
+    memcpy(address_n, _address_n, sizeof(address_n));
+    address_n_count = _address_n_count;
+
+    int n;
+    sha256_Init(&ctx);
+    char buffer[SHA256_BLOCK_LENGTH + 1]; // NULL TERMINATOR NOT PART OF HASH
+
+    n = snprintf(buffer, SHA256_BLOCK_LENGTH + 1, SIGNING_TEMPLATE_START_SEG1, account_number);
+    if (n < 0) { return false; }
+    sha256_Update(&ctx, (uint8_t *)buffer, n);
+
+    sha256UpdateEscaped(&ctx, chain_id, chain_id_length);
+
+    n = snprintf(buffer, SHA256_BLOCK_LENGTH + 1, SIGNING_TEMPLATE_START_SEG2, fee_uatom_amount);
+    if (n < 0) { return false; }
+    sha256_Update(&ctx, (uint8_t *)buffer, n);
+
+    n = snprintf(buffer, SHA256_BLOCK_LENGTH + 1, SIGNING_TEMPLATE_START_SEG3, gas);
+    if (n < 0) { return false; }
+    sha256_Update(&ctx, (uint8_t *)buffer, n);
+
+    sha256UpdateEscaped(&ctx, memo, memo_length);
+
+    sha256_Update(&ctx, (uint8_t *)SIGNING_TEMPLATE_START_SEG4, sizeof(SIGNING_TEMPLATE_START_SEG4) - 1);
+
+    return true;
+}
+
+bool cosmos_signTxUpdateMsgSend(const uint64_t amount,
+                                const char *from_address,
+                                const char *to_address)
+{
+    int n;
+    char buffer[SHA256_BLOCK_LENGTH + 1]; // NULL TERMINATOR NOT PART OF HASH
+
+    size_t decoded_len;
+    char hrp[45];
+    uint8_t decoded[38];
+    if (!bech32_decode(hrp, decoded, &decoded_len, from_address)) { return false; }
```

Comment:
Can the `from_address` be derived from the signer's pubkey? When would it ever not be the same? If so, we should do address encoding here, instead of decode & check.

Reply:
Yes we can do that

Reply:
ok, in that case, it's fine as-is.